src/processor.ts

```diff
@@ -1,63 +1,74 @@
-import { parseNodeNatspec } from './parser';
-import { Config } from './utils';
-import { validate } from './validator';
-import { SourceUnit, FunctionDefinition } from 'solc-typed-ast';
 import fs from 'fs';
+import { Parser } from './parser';
+import { Config } from './types/config.t';
+import { Validator } from './validator';
+import { SourceUnit, FunctionDefinition, ContractDefinition } from 'solc-typed-ast';
+import { NodeToProcess } from './types/solc-typed-ast.t';
 
 interface IWarning {
   location: string;
   messages: string[];
 }
 
-export async function processSources(sourceUnits: SourceUnit[], config: Config): Promise<IWarning[]> {
-  let warnings: IWarning[] = [];
-
-  sourceUnits.forEach((sourceUnit) => {
-    sourceUnit.vContracts.forEach((contract) => {
-      [
-        ...contract.vEnums,
-        ...contract.vErrors,
-        ...contract.vEvents,
-        ...contract.vFunctions,
-        ...contract.vModifiers,
-        ...contract.vStateVariables,
-        ...contract.vStructs,
-      ].forEach((node) => {
-        if (!node) return;
-
-        const nodeNatspec = parseNodeNatspec(node);
-        const validationMessages = validate(node, nodeNatspec, config);
-
-        // the constructor function definition does not have a name, but it has kind: 'constructor'
-        const nodeName = node instanceof FunctionDefinition ? node.name || node.kind : node.name;
-        const sourceCode = fs.readFileSync(sourceUnit.absolutePath, 'utf8');
-        const line = lineNumber(nodeName as string, sourceCode);
-
-        if (validationMessages.length) {
-          warnings.push({
-            location: `${sourceUnit.absolutePath}:${line}\n${contract.name}:${nodeName}`,
-            messages: validationMessages,
-          });
-        }
-      });
-    });
-  });
-
-  return warnings;
-}
+export class Processor {
+  config: Config;
+  validator: Validator;
+  parser: Parser;
 
-function lineNumberByIndex(index: number, string: string): Number {
-  let line = 0;
-  let match;
-  let re = /(^)[\S\s]/gm;
+  constructor(config: Config) {
+    this.config = config;
+    this.validator = new Validator(config);
+    this.parser = new Parser();
+  }
 
-  while ((match = re.exec(string))) {
-    if (match.index > index) break;
-    line++;
+  processSources(sourceUnits: SourceUnit[]): IWarning[] {
+    return sourceUnits.flatMap((sourceUnit) =>
+      sourceUnit.vContracts.flatMap((contract) =>
+        this.selectEligibleNodes(contract).flatMap((node) => this.validateNodeNatspec(sourceUnit, node, contract))
```

Comment:
u could use map instead of flatMap in this one

Reply:
In this case, `flatMap` is a way to filter out empty warnings. `[[], []].flat()` => `[]`.
But if we change it to return an `IWarning` from `validateNodeNatspec`, we will have to call filter (or something else) instead of `flatMap` and it will be way more verbose.